File: languages/java/app-encryption/src/test/java/com/godaddy/asherah/appencryption/persistence/InMemoryMetastoreImplTest.java
package com.godaddy.asherah.appencryption.persistence;

import static org.junit.jupiter.api.Assertions.*;

import java.time.Instant;
import java.time.temporal.ChronoUnit;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

class InMemoryMetastoreImplTest {
  InMemoryMetastoreImpl<String> inMemoryMetastoreImpl;

  @BeforeEach
  void setUp() {
    inMemoryMetastoreImpl = new InMemoryMetastoreImpl<>();
  }

  @Test
  void testLoadAndStoreWithValidKey() {
    String keyId = "ThisIsMyKey";
    Instant created = Instant.now();
    String value = "This is my value";

    inMemoryMetastoreImpl.store(keyId, created, value);

    String retrievedValue = inMemoryMetastoreImpl.load(keyId, created).get();

    assertEquals(value, retrievedValue, "Value loaded from persistence doesn't match what was stored.");
  }

  @Test
  void testLoadAndStoreWithInvalidKey() {
    String keyId = "ThisIsMyKey";
    Instant created = Instant.now();
    String value = "This is my value";

    inMemoryMetastoreImpl.store(keyId, created, value);

    assertFalse(inMemoryMetastoreImpl.load("some non-existent key", created).isPresent(),
        "Found value for non-existent key");
  }

  @Test
  void testLoadLatestMultipleCreatedAndValuesForKeyIdShouldReturnLatest() {
    String keyId = "ThisIsMyKey";
    Instant created = Instant.now();
    String value = "created value";
    inMemoryMetastoreImpl.store(keyId, created, value);

    Instant createdOneHourLater = created.plus(1, ChronoUnit.HOURS);
    String valueCreatedOneHourLater = value + createdOneHourLater;
    inMemoryMetastoreImpl.store(keyId, createdOneHourLater, valueCreatedOneHourLater);

    Instant createdOneDayLater = created.plus(1, ChronoUnit.DAYS);
    String valueCreatedOneDayLater = value + createdOneDayLater;
    inMemoryMetastoreImpl.store(keyId, createdOneDayLater, valueCreatedOneDayLater);

    Instant createdOneWeekEarlier = created.minus(7, ChronoUnit.DAYS);
    String valueCreatedOneWeekEarlier = value + createdOneWeekEarlier;
    inMemoryMetastoreImpl.store(keyId, createdOneWeekEarlier, valueCreatedOneWeekEarlier);

    assertEquals(valueCreatedOneDayLater, inMemoryMetastoreImpl.loadLatest(keyId).get());
  }

  @Test
  void testLoadLatestNonExistentKeyIdShouldReturnNull() {
    String keyId = "ThisIsMyKey";
    Instant created = Instant.now();
    String value = "created value";
    inMemoryMetastoreImpl.store(keyId, created, value);

    assertFalse(inMemoryMetastoreImpl.loadLatest("some non-existent key").isPresent());
  }

  @Test
  void testStoreWithDuplicateKeyShouldReturnFalse() {
    String keyId = "ThisIsMyKey";
    Instant created = Instant.now();
    String value = "this is the value";

    assertTrue(inMemoryMetastoreImpl.store(keyId, created, value));
    assertFalse(inMemoryMetastoreImpl.store(keyId, created, value));
  }
}

File: csharp/AppEncryption/AppEncryption.Tests/AppEncryption/Persistence/InMemoryMetastoreImplTest.cs
using System;
using GoDaddy.Asherah.AppEncryption.Persistence;
using LanguageExt;
using Xunit;

namespace GoDaddy.Asherah.AppEncryption.Tests.AppEncryption.Persistence
{
    [Collection("Logger Fixture collection")]
    public class InMemoryMetastoreImplTest
    {
        private readonly InMemoryMetastoreImpl<string> inMemoryMetastoreImpl;

        public InMemoryMetastoreImplTest()
        {
            inMemoryMetastoreImpl = new InMemoryMetastoreImpl<string>();
        }

        [Fact]
        private void TestLoadAndStoreWithValidKey()
        {
            const string keyId = "ThisIsMyKey";
            DateTimeOffset created = DateTimeOffset.UtcNow;
            const string value = "This is my value";

            inMemoryMetastoreImpl.Store(keyId, created, value);

            Option<string> actualValue = inMemoryMetastoreImpl.Load(keyId, created);

            Assert.Equal(value, actualValue);
        }

        [Fact]
        private void TestLoadAndStoreWithInvalidKey()
        {
            const string keyId = "ThisIsMyKey";
            DateTimeOffset created = DateTimeOffset.UtcNow;
            const string value = "This is my value";

            inMemoryMetastoreImpl.Store(keyId, created, value);

            Option<string> actualValue = inMemoryMetastoreImpl.Load("some non-existent key", created);

            Assert.True(actualValue.IsNone);
        }

        [Fact]
        private void TestLoadLatestMultipleCreatedAndValuesForKeyIdShouldReturnLatest()
        {
            const string keyId = "ThisIsMyKey";
            DateTimeOffset created = DateTimeOffset.UtcNow;
            const string value = "This is my value";

            inMemoryMetastoreImpl.Store(keyId, created, value);

            DateTimeOffset createdOneHourLater = created.AddHours(1);
            string valueCreatedOneHourLater = value + createdOneHourLater;
            inMemoryMetastoreImpl.Store(keyId, createdOneHourLater, valueCreatedOneHourLater);

            DateTimeOffset createdOneDayLater = created.AddDays(1);
            string valueCreatedOneDayLater = value + createdOneDayLater;
            inMemoryMetastoreImpl.Store(keyId, createdOneDayLater, valueCreatedOneDayLater);

            DateTimeOffset createdOneWeekEarlier = created.AddDays(-7);
            string valueCreatedOneWeekEarlier = value + createdOneWeekEarlier;
            inMemoryMetastoreImpl.Store(keyId, createdOneWeekEarlier, valueCreatedOneWeekEarlier);

            Option<string> loadLatest = inMemoryMetastoreImpl.LoadLatest(keyId);

            Assert.Equal(valueCreatedOneDayLater, loadLatest);
        }

        [Fact]
        private void TestLoadLatestNonExistentKeyIdShouldReturnNull()
        {
            const string keyId = "ThisIsMyKey";
            DateTimeOffset created = DateTimeOffset.UtcNow;
            const string value = "This is my value";

            inMemoryMetastoreImpl.Store(keyId, created, value);

            Assert.True(inMemoryMetastoreImpl.LoadLatest("some non-existent key").IsNone);
        }

        [Fact]
        private void TestStoreWithDuplicateKeyShouldReturnFalse()
        {
            const string keyId = "ThisIsMyKey";
            DateTimeOffset created = DateTimeOffset.UtcNow;
            const string value = "This is my value";

            Assert.True(inMemoryMetastoreImpl.Store(keyId, created, value));
            Assert.False(inMemoryMetastoreImpl.Store(keyId, created, value));
        }
    }
}
